js/SheetsApi.js

```diff
@@ -0,0 +1,204 @@
+     * @returns       a 2D array
+     */
+    function filterByKeyword(values, keyword, columnIndex) {
+        let shouldStay = false;
+        let rows = values.length;
+        for (let i = 0; i < rows; i++) {
+            shouldStay = false;
+            if (columnIndex < 0) {
+                for (let j = 0; j < values[i].length; j++) {
+                    if (values[i][j].includes(keyword)) {
+                        shouldStay = true;
+                        break;
+                    }
+                }
+            } else {
+                if (values[i][columnIndex].includes(keyword)) {
+                    shouldStay = true;
+                }
+            }
+            if (!shouldStay) {
+                values.splice(i, 1);
+                rows--;
+                i--;
+            }
+        }
+        return values;
+    }
+
+    /**
+     * Returns a promise for adding the inputValues to the sheet
+     * @param inputRange   a string value of target sheet in the spreadsheet
+     * @param inputValues a 2D array of the values to be added
+     * @returns           a promise
+     */
+    function update(inputRange, inputValues) {
+        let params = {
+            spreadsheetId: sheetId,
+            range: inputRange,
+            valueInputOption: "USER_ENTERED",
+            values: inputValues
+        }
+        return gapi.client.sheets.spreadsheets.values.update(params);
+    }
+
+    /**
+     * This method returns the cells updated of batchAdd
+     * @param response the response of update(inputRange, inputValues)
+     * @return number of rows updated
+     */
+    function parseUpdate(response) {
+        return response.result.updatedCells;
+    }
+
+    /**
+     * Returns a promise for adding the inputValues to the sheet
+     * @param inputRange   a string value of target sheet in the spreadsheet
+     * @param inputValues a 2D array of the values to be added
+     * @returns           a promise
+     */
+    function batchAdd(inputRange, inputValues) {
+        let params = {
+            spreadsheetId: sheetId,
+            range: inputRange + "!A:A",
+            majorDimension: "ROWS",
+            valueInputOption: "USER_ENTERED",
+            values: inputValues
+        };
+        return gapi.client.sheets.spreadsheets.values.append(params);
+    }
+
+    /**
+     * This method returns the rows updated of batchAdd
+     * @param response the response of batchAdd(inputRange, inputValues)
+     * @return number of rows updated
+     */
+    function parseBatchAdd(response) {
+        return response.result.updates.updatedRows;
+    }
+
+    /**
+     * This function parses the error response and logs the error message
+     * to the console.error and returns the error message. 
+     * @param reason the error response of a promise
+     */
+    function parseErrorMessage(reason) {
+        console.error('error: ' + reason.result.error.message);
+        return reason.result.error.message;
+    }
+
```

Comment:
Function to add:
Essentially an SQL select statement: SELECT return_cols FROM sheet_name WHERE condition1 AND condition2 AND ...etc.

ORs will not often be useful in this system, so they can be done hackily if need-be by just calling this method for each condition that needs to be ORed together as that will give the same result.

selectFromTableWhereConditions(array return_cols, string sheet_name, array conditions)
With the conditions are organized as follows:
conditions = [condition1, condition2....etc]
condition1 = {header:"the name of a header", value:"the value to check for"}
If the values of any param are a '*', then we treat it as a wildcard just like SQL.
The return value should be an array of js objects with the following format:
return = [
  {header1:"value1_1", header2:"value1_2"...}
  {header1:"value2_1", header2:"value2_2"...}
  and so on for each row returned.
]

More SQL functions to add:
SQL--> INSERT INTO sheet_name (header1, header2,...etc) VALUES (value1, value2,...etc)
JS--> insertIntoTableColValues(string sheet_name, object to_insert)
to_insert would be {header1:"value1", header2:"value2",...etc}
SQL--> UPDATE sheet_name SET header1 = value1, header2 = value2, ...etc WHERE conditions
JS--> updateTable(string sheet_name, object col_val, object conditions)
col_val would be {header1:"value1", header2:"value2",...etc}
conditions would be the same as previously specified in the SELECT suggestion.
SQL--> ALTER TABLE sheet_name ADD header datatype
JS--> alterTableAddCol(string sheet_name, string header, string datatype)

Also a function to get header names, mostly for getting a list of categories:
JS--> getColsFromTable(string sheet_name)
returns an array of headers

Reply:
for the alterTableAddCol, the datatype is not available. The data is either USER_ENTERED or RAW. See this https://developers.google.com/sheets/api/reference/rest/v4/ValueInputOption.
The RAW can be a solution for the "0123456" and "123456" problem, and the only disadvantage is that data will stay as strings, which in my perspective is not a problem.

Reply:
It looks like USER_ENTERED would also work, because that seems like it would automatically strip out leading zeros?

Reply:
I changed the Select function. It now needs all values from getSheet(sheetName) and then applies the SELECT colNames WHERE conditions. It is better to use RAW so the the values fetched from getSheet will have the leading 0 and then conditions having leading 0 or not doesn't matter since my searching is based on String.contains()

Reply:
Sounds good!